ros_ign_gazebo/src/optical_frame_publisher.cpp

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright (C) 2022 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "ros_ign_gazebo/optical_frame_publisher.hpp"
+
+#include <tf2/LinearMath/Quaternion.h>
+#include <tf2_ros/transform_broadcaster.h>
+#include <tf2_ros/static_transform_broadcaster.h>
+
+#include <memory>
+#include <string>
+
+#include <geometry_msgs/msg/transform_stamped.hpp>
+#include <sensor_msgs/msg/image.hpp>
+#include <sensor_msgs/msg/camera_info.hpp>
+
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp_components/register_node_macro.hpp>
+
+using std::placeholders::_1;
+using namespace std::chrono_literals;
+
+namespace ros_ign_gazebo
+{
+struct OpticalFramePublisher::Impl
+{
+  /// \brief Check for downstream subscriptions and enable upstream
+  /// subscriptions based on that
+  void CheckSubscribers();
+
+  /// \brief Start upstream subscription for images
+  void ImageConnect();
+
+  /// \brief Start upstream subscription for camera info
+  void CameraInfoConnect();
+
+  /// \brief Image callback
+  void UpdateImageFrame(const sensor_msgs::msg::Image & msg);
+
+  /// \brief Camera info callback
+  void UpdateCameraInfoFrame(const sensor_msgs::msg::CameraInfo & msg);
+
+  /// \brief Publish new TF between frame and optical_frame
+  void PublishTF(const std::string & frame, const std::string & child_frame);
+
+  /// \brief Interface for creating publications/subscriptions
+  rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr node_topics;
+
+  /// \brief timer with callback to check for publisher subscription count
+  rclcpp::TimerBase::SharedPtr timer;
+
+  /// \brief Static transform broadcaster that broadcasts the optical frame id
+  std::unique_ptr<tf2_ros::StaticTransformBroadcaster> tfBroadcasterStatic;
```

Comment:
Should we unify the naming of the member variables (camel or snake)?

Reply:
Changed to snake case, since that's what it mostly was